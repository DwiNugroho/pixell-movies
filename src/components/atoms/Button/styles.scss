@mixin button-mixin($text, $border, $bg) {
  color: $text;
  border: 1px solid $border;
  background-color: $bg;
  &:hover {
    border: 1px solid darken($color: $border, $amount: 4%);
    background-color: darken($color: $bg, $amount: 4%);
  }
  &:disabled,
  &[disabled] {
    color: lighten($color: $text, $amount: 16%);
    border: 1px solid lighten($color: $border, $amount: 16%);
    background-color: lighten($color: $bg, $amount: 16%);
  }
}

.atom-button {
  cursor: pointer;
  outline: none;
  position: relative;
  overflow: hidden;
  border-radius: 4px;
  font-weight: 600 !important;
  transition: 0.4s ease-in-out;
  &:disabled,
  &[disabled] {
    cursor: not-allowed;
  }

  &--small {
    padding: 2px 8px;
  }
  &--medium {
    padding: 4px 12px;
  }
  &--large {
    padding: 8px 16px;
  }

  &--block {
    width: 100%;
  }

  &--primary {
    color: $light-black;
    border: 1px solid $quill-gray;
    background-color: $quill-gray;
    &:disabled,
    &[disabled] {
      color: lighten($color: $light-black, $amount: 16%);
      border: 1px solid lighten($color: $quill-gray, $amount: 4%);
    }
    &:hover {
      border: 1px solid darken($color: $quill-gray, $amount: 4%);
      background-color: darken($color: $quill-gray, $amount: 4%);
    }
    &.atom-button--blue {
      @include button-mixin($white, $dark-blue, $blue);
    }
    &.atom-button--red {
      @include button-mixin($white, $dark-red, $red);
    }
    &.atom-button--yellow {
      @include button-mixin($white, $dark-yellow, $yellow);
    }
  }

  &--outline {
    color: $light-black;
    border: 1px solid $quill-gray;
    background-color: transparent;
    &:disabled,
    &[disabled] {
      color: lighten($color: $light-black, $amount: 16%);
      border: 1px solid lighten($color: $quill-gray, $amount: 4%);
      &:hover {
        background-color: transparent;
      }
    }
    &:hover {
      border: 1px solid darken($color: $quill-gray, $amount: 4%);
      background-color: rgba($color: #000000, $alpha: 0.04);
    }

    &.atom-button--blue {
      @include button-mixin($blue, $blue, $lighter-blue);
    }
    &.atom-button--red {
      @include button-mixin($red, $red, $lighter-red);
    }
    &.atom-button--yellow {
      @include button-mixin($yellow, $yellow, $lighter-yellow);
    }
  }

  &__content {
    position: relative;
    z-index: 2;
  }

  &__ripple {
    width: 20px;
    height: 20px;
    position: absolute;
    background: rgba($color: #000000, $alpha: 0.2);
    display: block;
    content: '';
    border-radius: 9999px;
    opacity: 1;
    animation: 1.6s ease 1 forwards ripple-effect;
  }
}
