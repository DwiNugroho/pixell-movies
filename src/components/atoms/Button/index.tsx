import * as React from 'react';

import './styles.scss';

export interface Props extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /**
   * Button contents
   */
  children: React.ReactNode;
  /**
   * Optional click handler
   */
  onClick?: () => void;
  /**
   *  Spans the full width of the Button parent
   */
  block?: boolean;
  /**
   * A button can have different sizes
   */
  size?: 'small' | 'medium' | 'large';
  /**
   *  A button can have different appearances
   */
  appearance?: 'primary' | 'outline';
  /**
   * A button can have different colors
   */
  color?: 'none' | 'red' | 'blue' | 'yellow';
  /**
   * A button can show it is currently unable to be interacted with
   */
  disabled?: boolean;
  /**
   *  A button can have className atrribute
   */
  className?: string;
}

/**
 * Reusable Button Component
 */
export const Button: React.FC<Props> = ({
  children,
  onClick,
  size,
  appearance,
  color,
  disabled,
  block,
  className,
  ...props
}) => {
  const [coords, setCoords] = React.useState({ x: -1, y: -1 });
  const [isRippling, setIsRippling] = React.useState(false);

  React.useEffect(() => {
    if (coords.x !== -1 && coords.y !== -1) {
      setIsRippling(true);
      setTimeout(() => setIsRippling(false), 400);
    } else setIsRippling(false);
  }, [coords]);

  React.useEffect(() => {
    if (!isRippling) setCoords({ x: -1, y: -1 });
  }, [isRippling]);

  const getSize =
    (size || '').toLocaleLowerCase() === 'large'
      ? 'atom-button--large'
      : (size || '').toLocaleLowerCase() === 'small'
      ? 'atom-button--small'
      : 'atom-button--medium';

  const getApperiance =
    (appearance || '').toLocaleLowerCase() === 'outline'
      ? 'atom-button--outline'
      : 'atom-button--primary';

  const getColor =
    (color || '').toLocaleLowerCase() === 'red'
      ? 'atom-button--red'
      : (color || '').toLocaleLowerCase() === 'blue'
      ? 'atom-button--blue'
      : (color || '').toLocaleLowerCase() === 'yellow'
      ? 'atom-button--yellow'
      : '';

  const isBlock = block ? 'atom-button--block' : '';

  const haveClassName = className ? className : '';

  return (
    <button
      type="button"
      className={`atom-button ${getSize} ${getApperiance} ${getColor} ${isBlock} ${haveClassName}`}
      onClick={(e) => {
        const rect = (e.target as any).getBoundingClientRect();
        setCoords({ x: e.clientX - rect.left, y: e.clientY - rect.top });
        onClick && onClick();
      }}
      disabled={disabled}
      {...props}
    >
      {isRippling ? (
        <span
          className="atom-button__ripple"
          style={{
            left: coords.x,
            top: coords.y,
          }}
        />
      ) : (
        ''
      )}
      {children}
    </button>
  );
};

Button.defaultProps = {
  children: 'Button',
  onClick: undefined,
  size: 'medium',
  appearance: 'primary',
  color: 'none',
};

export default Button;
