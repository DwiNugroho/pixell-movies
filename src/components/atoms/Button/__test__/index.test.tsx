import React from 'react';
import ReactDOM from 'react-dom';
import ReactTestUtils from 'react-dom/test-utils';
import { render, fireEvent } from '@testing-library/react';
import Button from '../index';

test('renders without crashing', () => {
  const section = document.createElement('section');
  ReactDOM.render(<Button>Text</Button>, section);
});

test('Should show the default type', () => {
  const { container } = render(<Button>Text</Button>);
  const button = container.querySelector('button');

  expect(button).toHaveAttribute('type', 'button');
});

test('should show the submit type', () => {
  const { container } = render(<Button type="submit">Text</Button>);
  const button = container.querySelector('button');

  expect(button).toHaveAttribute('type', 'submit');
});

test('Should be disabled', () => {
  const { container } = render(<Button disabled>Text</Button>);
  const button = container.querySelector('button');

  expect(button).toBeDisabled();
});

test('Should have a custom className', () => {
  const { container } = render(<Button className="custom">Text</Button>);
  const button = container.querySelector('button');

  expect(button).toHaveClass('custom');
});

test('Should have a custom style', () => {
  const { container } = render(
    <Button style={{ fontSize: '12px' }}>Text</Button>,
  );
  const button = container.querySelector('button');

  expect(button).toHaveStyle('font-size: 12px');
});

test('Should have block class', () => {
  const { container } = render(<Button block>Text</Button>);
  const button = container.querySelector('button');

  expect(button).toHaveClass('atom-button--block');
});

test('Should apply appearance', () => {
  const { container } = render(<Button appearance="outline">Text</Button>);
  const button = container.querySelector('button');

  expect(button?.className).toMatch(/outline/i);
});

test('Should apply color', () => {
  const { container } = render(<Button color="blue">Text</Button>);
  const button = container.querySelector('button');

  expect(button?.className).toMatch(/blue/i);
});

test('Should apply size', () => {
  const { container } = render(<Button size="medium">Text</Button>);
  const button = container.querySelector('button');

  expect(button?.className).toMatch(/medium/i);
});

test('Should honour additional classes passed in, adding not overriding', () => {
  const { container } = render(
    <Button className="button-custom" size="medium">
      Text
    </Button>,
  );
  const button = container.querySelector('button');

  expect(button).toHaveClass('button-custom');
  expect(button?.className).toMatch(/medium/i);
});

test('Should call onClick callback', (done) => {
  const clickHandler = () => {
    done();
  };

  const { container } = render(
    <Button
      onClick={() => {
        clickHandler();
      }}
    >
      Text
    </Button>,
  );
  const button = container.querySelector('button') || container;

  ReactTestUtils.Simulate.click(button);
});
