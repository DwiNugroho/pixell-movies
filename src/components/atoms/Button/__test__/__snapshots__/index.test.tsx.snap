// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ripple 1`] = `
<DocumentFragment>
  <button
    class="atom-button atom-button--medium atom-button--primary   "
    type="button"
  >
    <span
      class="atom-button__ripple"
      style="left: 0px; top: 0px;"
    />
    Text
  </button>
</DocumentFragment>
`;
